pkg/kubelet/network/hairpin/hairpin.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hairpin
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+const (
+	sysfsNetPath            = "/sys/devices/virtual/net"
+	hairpinModeRelativePath = "brport/hairpin_mode"
+	hairpinEnable           = "1"
+)
+
+func SetUpContainer(containerPid int, containerInterfaceName string) error {
+	hostIfName, err := FindPairInterfaceOfContainerInterface(containerPid, containerInterfaceName)
+	if err != nil {
+		return err
+	}
+	return SetUpInterface(hostIfName)
+}
+
+// This may be moved to a util package.
+func FindPairInterfaceOfContainerInterface(containerPid int, containerInterfaceName string) (string, error) {
+	// Get container's interface index
+	output, err := exec.Command("nsenter", "-t", fmt.Sprintf("%d", containerPid), "-n", "-F", "ip", "link", "show", containerInterfaceName).Output()
+	if err != nil {
+		return "", fmt.Errorf("Unable to query interface %s of container %d: %v", containerInterfaceName, containerPid, err)
+	}
+	// output is like "index: informations"
+	containerIfindex := strings.TrimSpace(strings.SplitN(string(output), ":", 2)[0])
+
+	// Lookup the matching host interface (if we could reliably get the iflink
+	// value for container's interface, we could use net.InterfaceByIndex(...)) instead.
+	interfaces, err := net.Interfaces()
+	for _, netIf := range interfaces {
+		iflink, err := ioutil.ReadFile(path.Join(sysfsNetPath, netIf.Name, "iflink"))
+		if err != nil {
+			continue
+		}
+		if strings.TrimSpace(string(iflink)) == containerIfindex {
```

Comment:
Creating a veth device pair may be atomic in the kernel, but I prefered to use iflink instead of relying on ifindex+1 or -1 (depending on the side), because it is semantic. We may try to use this as an heuristic to optimize the loop, but it is more code complexity.